keras_rcnn/preprocessing/_object_detection.py

```diff
@@ -1,118 +1,84 @@
+        scale = max_size / max_shape
 
-                self.batch_index += 1
-            else:
-                current_batch_size = n - current_index
+    return (int(shape[0] * scale), int(shape[1] * scale), shape[2]), scale
 
-                self.batch_index = 0
 
-            self.total_batches_seen += 1
+def scale_image(image, min_size=224, max_size=224):
+    """
+    Rescales an image according to the heuristics from 'scale_shape'.
+    """
+    target_shape, scale = scale_shape(image.shape, min_size, max_size)
 
-            yield index_array[current_index:current_index + current_batch_size], current_index, current_batch_size
+    return skimage.transform.rescale(image, scale=scale, mode="reflect"), scale
 
-    def __iter__(self):
-        return self
 
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
+    def __init__(self, data, classes, image_data_generator, image_shape, batch_size=1,
+                 shuffle=True, seed=numpy.uint32(time.time() * 1000)):
+        self.data                 = data
+        self.classes              = classes
+        self.image_data_generator = image_data_generator
+        self.image_shape          = image_shape
+        self.target_shape, _      = scale_shape(image_shape)
 
-    def next(self, *args, **kwargs):
-        pass
-
-
-class DictionaryIterator(Iterator):
-    def __init__(self, dictionary, generator, shuffle=False, seed=None):
-        self.dictionary = dictionary
-
-        self.encoder = sklearn.preprocessing.LabelEncoder()
-
-        self.generator = generator
-
-        self.encoder.fit(generator.classes)
-
-        Iterator.__init__(self, len(dictionary), 1, shuffle, seed)
+        super().__init__(len(self.data), batch_size, shuffle, seed)
 
     def next(self):
+        # Lock indexing to prevent race conditions
         with self.lock:
-            index_array, current_index, current_batch_size = next(
-                self.index_generator)
-
-        index = index_array[0]
+            selection, _, batch_size = next(self.index_generator)
 
-        pathname = self.dictionary[index]["filename"]
+        # Transformation of images is not under thread lock so it can be done in parallel
+        image_batch    = numpy.zeros((batch_size,) + self.target_shape, dtype=keras.backend.floatx())
+        gt_boxes_batch = numpy.zeros((batch_size, 0, 5),                dtype=keras.backend.floatx())
+        metadata       = numpy.zeros((batch_size, 3),                   dtype=keras.backend.floatx())
 
-        image = skimage.io.imread(pathname)
+        for batch_index, image_index in enumerate(selection):
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
+            image, scale = scale_image(image)
+            image        = self.image_data_generator.random_transform(image)
```

Comment:
I would drop the `random_transform` and `standardize` calls for the moment.

Unfortunately, we can’t easily inherit from `ImageDataGenerator` since we’ll need to write augmentation and pre-processing functions that can manage images and bounding boxes together. We’ll also need to support masks in the near future.